o1vm/src/mips/interpreter.rs

```diff
@@ -501,6 +501,44 @@ pub trait InterpreterEnv {
         };
     }
 
+    /// Adds a lookup to the RangeCheck16Lookup table
+    fn lookup_16bits(&mut self, value: &Self::Variable) {
+        self.add_lookup(Lookup::read_one(
+            LookupTableIDs::RangeCheck16Lookup,
+            vec![value.clone()],
+        ));
+    }
+
+    /// Range checks with 2 lookups to the RangeCheck16Lookup table that a value
+    /// is at most 2^`bits`-1  (bits <= 16).
+    fn range_check16(&mut self, value: &Self::Variable, bits: u32) {
+        assert!(bits <= 16);
+        // 0 <= value < 2^bits
+        // First, check lowerbound: 0 <= value < 2^16
+        self.lookup_16bits(value);
+        // Second, check upperbound: value + 2^16 - 2^bits < 2^16
+        self.lookup_16bits(&(value.clone() + Self::constant(1 << 16) - Self::constant(1 << bits)));
+    }
+
+    /// Adds a lookup to the ByteLookup table
+    fn lookup_8bits(&mut self, value: &Self::Variable) {
+        self.add_lookup(Lookup::read_one(
+            LookupTableIDs::ByteLookup,
+            vec![value.clone()],
+        ));
+    }
+
+    /// Range checks with 2 lookups to the ByteLookup table that a value
+    /// is at most 2^`bits`-1  (bits <= 8).
+    fn range_check8(&mut self, value: &Self::Variable, bits: u32) {
+        assert!(bits <= 8);
+        // 0 <= value < 2^bits
+        // First, check lowerbound: 0 <= value < 2^8
+        self.lookup_8bits(value);
+        // Second, check upperbound: value + 2^8 - 2^bits < 2^8
+        self.lookup_8bits(&(value.clone() + Self::constant(1 << 8) - Self::constant(1 << bits)));
```

Comment:
Interesting approach. It does save an additional table and an additional multiplicity polynomial IIUC.